File: Assets/Scripts/Codex/ActiveEffect.cs
using System.Collections.Generic;

namespace UnmappedIsland.Codex
{
    /// <summary>
    /// 一時的な命令の内容（`set`/`add`/`destroy`/`spawn`、9節）。on_min・on_overflow・on_shortfall（6節）、
    /// actions/combinations/pickのactive（11・12・10節）のすべてから、この1つの型をそのまま共用する。
    ///
    /// 文法は「操作が上位、対象(self/parent/actor/dragged)が下位」（例: `add: {self: {hour: 1}}`）。
    /// set/addは複数の対象へ同時に書けるため、対象ごとの一覧を辞書として持つ。destroyは削除対象を
    /// 直接指すリスト（`destroy: self`は要素1つのリストとして表す）。spawnは常にselfが実行するものと
    /// みなすため対象キーを持たない（対象別に分かれているのはSpawn.Into＝配置先の起点であり、
    /// 実行者＝selfとは別の概念）。
    ///
    /// on_min/on_overflow/on_shortfallはselfのみが有効な対象（呼び出し側・パーサ側で強制する）。
    /// </summary>
    public sealed class ActiveEffect
    {
        /// <summary>対象ごとのset(絶対値代入)。空なら該当対象へのsetなし。</summary>
        public IReadOnlyDictionary<ReferenceRoot, IReadOnlyList<PropertyAssignment>> Sets { get; }

        /// <summary>対象ごとのadd(加減算)。空なら該当対象へのaddなし。</summary>
        public IReadOnlyDictionary<ReferenceRoot, IReadOnlyList<PropertyDelta>> Adds { get; }

        /// <summary>削除する対象。空なら destroy なし。</summary>
        public IReadOnlyList<ReferenceRoot> Destroy { get; }

        /// <summary>spawn。null なら spawn なし。常にselfが実行する。</summary>
        public SpawnEffect Spawn { get; }

        /// <summary>transfer（9.5節）。null なら transfer なし。fromとtoの組が常に1組のため、spawnと同じく
        /// 対象キーを持たない単一の値として持つ（`transfer`は複数形キーではないため、8節の
        /// 「複数形キーは常に配列」規約の対象外）。</summary>
        public TransferEffect Transfer { get; }

        public ActiveEffect(
            IReadOnlyDictionary<ReferenceRoot, IReadOnlyList<PropertyAssignment>> sets,
            IReadOnlyDictionary<ReferenceRoot, IReadOnlyList<PropertyDelta>> adds,
            IReadOnlyList<ReferenceRoot> destroy,
            SpawnEffect spawn,
            TransferEffect transfer = null)
        {
            Sets = sets;
            Adds = adds;
            Destroy = destroy;
            Spawn = spawn;
            Transfer = transfer;
        }
    }

    /// <summary>
    /// set の1エントリ（対象プロパティのグローバルIDと代入する絶対値）。ValueRefが非nullの場合、Valueの
    /// 代わりに、その{object, prop}参照先の現在の実効値を代入する（他のプロパティの値をそのままコピーする、
    /// conditionsのvalue参照・weightのpath参照と同じ「リテラルか参照か」の二択、9.2節）。
    /// </summary>
    public readonly struct PropertyAssignment
    {
        public readonly int PropertyGlobalId;
        public readonly int Value;
        public readonly PropertyPath? ValueRef;

        public PropertyAssignment(int propertyGlobalId, int value)
        {
            PropertyGlobalId = propertyGlobalId;
            Value = value;
            ValueRef = null;
        }

        public PropertyAssignment(int propertyGlobalId, PropertyPath valueRef)
        {
            PropertyGlobalId = propertyGlobalId;
            Value = default;
            ValueRef = valueRef;
        }
    }

    /// <summary>add の1エントリ（対象プロパティのグローバルIDと加減算量）。</summary>
    public readonly struct PropertyDelta
    {
        public readonly int PropertyGlobalId;
        public readonly int Amount;

        public PropertyDelta(int propertyGlobalId, int amount)
        {
            PropertyGlobalId = propertyGlobalId;
            Amount = amount;
        }
    }

    /// <summary>
    /// spawn の配置先（9.4節）が起点にする参照ルート。スロットは指定しない。対象オブジェクトが持つ
    /// スロットを宣言順に走査し、最初に配置できたスロットへ入れる（型ごとに用意されたスロットへ
    /// 自然に振り分けられるため、著者がスロット名を知っている必要がない）。
    ///
    /// fallback はYAML上に存在しない。配置に失敗した場合は必ず、解決した起点自身の親へ伝播する
    /// （WorldObject.Place参照）。Actor はアクション実行文脈でのみ解決できる。on_min/on_overflow/
    /// on_shortfallにはactorが存在しないため、それらのspawnでintoにActorを指定しても何も起きない。
    /// </summary>
    public enum SpawnTargetRoot
    {
        /// <summary>
        /// into を省略した場合の既定値でもある。この spawn を宣言したオブジェクト（self）が今いる、
        /// まさにその場所（親と、self が現在占めているのと同じスロット）へ配置する。クラフト・腐敗など
        /// 「同じ場所で別の物に置き換わる」場合に使う。一意に決まる1つのスロットのため、走査は行わない。
        /// </summary>
        SameSlot,

        /// <summary>self が持つスロットを宣言順に走査する。</summary>
        Self,

        /// <summary>actor が持つスロットを宣言順に走査する。</summary>
        Actor,
    }

    /// <summary>
    /// spawn（9.4節）の内容。Into への配置（起点が持つスロットの宣言順走査、または SameSlot の場合は
    /// 一意に決まる1スロットへの直接配置）に失敗した場合、必ずその起点自身の親へ伝播し、
    /// accepts/capacityを無視して強制的に配置する（すべてのオブジェクトは必ずどこかの親に属さなければ
    /// ならないため）。この伝播はYAML側で選択の余地がなく、常に同じルールで行われる。
    ///
    /// 伝播先の親も存在しない場合（起点がworld直下など）、spawn したオブジェクトはどこにも配置されない
    /// まま消える（何も起きなかったのと同じ扱い）。
    /// </summary>
    public sealed class SpawnEffect
    {
        public int ObjectGlobalId { get; }

        public SpawnTargetRoot Into { get; }

        public SpawnEffect(int objectGlobalId, SpawnTargetRoot into)
        {
            ObjectGlobalId = objectGlobalId;
            Into = into;
        }
    }

    /// <summary>
    /// transfer（9.5節）の内容。fromが指すプロパティの実体値から、実際に出せる量とamountの小さい方だけを
    /// toが指すプロパティへ移す（連続量の液体のような、setの固定値でもaddの無条件加減算でも表現できない
    /// 「在庫に応じて実際に動く量が変わる」移送を表す）。
    ///
    /// from/toの参照は、他の場所（`modify`/`set`/`add`等）と同じ「対象がキー、内容が値」という規約に
    /// 合わせず、conditions（14節）と同じくフラットな`from_object`/`from_prop`/`to_object`/`to_prop`の
    /// 4フィールドで表す。fromとtoの組は常に1組であり、複数プロパティの入れ物としてネストする理由が
    /// 無いため。
    /// </summary>
    public sealed class TransferEffect
    {
        /// <summary>移送元の参照ルート。既定値self（conditionsのobject省略時と同じ規約）。</summary>
        public ReferenceRoot FromObject { get; }

        public int FromPropertyGlobalId { get; }

        /// <summary>移送先の参照ルート。既定値self。</summary>
        public ReferenceRoot ToObject { get; }

        public int ToPropertyGlobalId { get; }

        /// <summary>一度に移送を試みる量の上限。実際の移動量はこれとfromの残量（・allow_overflowが falseなら
        /// toの残容量）の小さい方になる。</summary>
        public int Amount { get; }

        /// <summary>
        /// falseの場合（既定）、toのrangeで実際に収まる量までしか移動しない（収まらない分はfromに残す。
        /// 液体を無駄にしない）。trueの場合、fromの残量とamountだけで移動量を決め、toのrange超過分は
        /// toのon_overflow（未指定ならrange.maxへ自動でクランプする既定動作）に委ねる（あふれた分は失われる）。
        /// </summary>
        public bool AllowOverflow { get; }

        public TransferEffect(
            ReferenceRoot fromObject, int fromPropertyGlobalId,
            ReferenceRoot toObject, int toPropertyGlobalId,
            int amount, bool allowOverflow)
        {
            FromObject = fromObject;
            FromPropertyGlobalId = fromPropertyGlobalId;
            ToObject = toObject;
            ToPropertyGlobalId = toPropertyGlobalId;
            Amount = amount;
            AllowOverflow = allowOverflow;
        }
    }
}

File: src/domain/ActiveEffect.ts
import type { InfluenceWriter } from './PropertyInfluence';
import type { PropertyValue } from './PropertyValue';
import type { SameSlotSpawnSite } from './SameSlotSpawnSite';
import type { WorldObject } from './WorldObject';
import type { WorldSession } from './WorldSession';
import type { ObjectRef } from './ObjectRef';
import type { AddReading, EffectReader, SetValueReading, TransferReading } from './EffectReader';
import type { PropertyPath, ReferenceContext } from './ReferenceRoot';

/**
 * 「条件成立時に何を起こすか」を表すポリモーフィックな効果1つ（9・10節）。対象の解決と適用まで自分で行う。
 * 具象は、単一の命令（Set/Add/Destroy/Spawn/Transfer、9節）、その宣言順合成（ActiveEffectSequence）、
 * weightで1候補を選ぶpick（PickEffect、10節。候補もActiveEffectなので再帰しうる）の3種。
 * pickは9節の命令と対等な1つの効果なので、合成の中に他の命令と並べて置ける。
 *
 * sameSlotSpawnSiteは、適用の入口（WorldObject.applyActiveEffect）で捕捉した「selfが今占めている位置」の
 * スナップショット。same_slot spawnだけがこれを使い、self破棄後でも「その位置がまだ同種を保持しているか」を
 * 配置時に見て置き換え位置を決める（他の効果は無視してよく、destroyが何かを書き込む必要もない）。
 */
export abstract class ActiveEffect {
  abstract apply(
    context: ReferenceContext,
    session: WorldSession,
    sameSlotSpawnSite: SameSlotSpawnSite | undefined,
  ): void;

  /**
   * この効果が何を宣言しているかを読み上げる（EffectReader参照）。**抽象なのは取りこぼしを防ぐため**
   * ——既定を持たせると、動詞を1つ足したときに読み手が黙って何も受け取らなくなる。
   */
  abstract read(reader: EffectReader): void;

  /**
   * まとめて実行するとき、回数の上限を決める器がいくつあるか（`allow_multiple`、12.4節）。
   * undefinedは「繰り返すと意味が変わるので数えられない」（pick）。
   *
   * **既定は0＝数に影響しない。** 取りこぼしても「まとめられない」に倒れるだけで安全側なので、readと
   * 違い抽象にしない。数を決められるのは、単調に埋まる器へ入れる効果（transfer）だけ。
   */
  repeatLimitingVesselCount(): number | undefined {
    return 0;
  }

  /**
   * 今この文脈で、この効果が行き先を持たないために、宣言している操作そのものが成立しないか
   * （`become` の座標が空、9.9節）。成立しない操作は候補に出さない——落とせるのに何も起きない、を
   * 作らないため。
   *
   * **既定はfalse＝妨げない。** repeatLimitingVesselCountと同じく、取りこぼしても安全側（操作は出る）に
   * 倒れるので抽象にしない。
   */
  blocksOperation(_context: ReferenceContext): boolean {
    return false;
  }

  /**
   * candidatesを先頭から順に、この効果を続けて何回適用できるか。undefinedは「答えられない」。
   * 各candidateはdraggedの役で、器（repeatLimitingVesselCount）を持つ効果だけが答える。
   */
  acceptedCount(_context: ReferenceContext, _candidates: readonly WorldObject[]): number | undefined {
    return undefined;
  }
}

/**
 * 一時的な命令（`set`/`add`/`destroy`/`spawn`/`transfer`、9節）と`pick`（10節）を、書かれた順に
 * まとめた合成効果。on_max・on_min（6節）、actions/combinations/pickの中身（11・12・10節）が
 * 共用する。on_max/on_minはselfのみが有効な対象（パーサ側で強制する）。
 * 空（命令が1つも無い）なら、適用しても何も起きない。
 */
export class ActiveEffectSequence extends ActiveEffect {
  /** 効果の宣言順リスト。適用順はリスト順で、パーサはYAMLに書かれた順のまま渡す（9.7節）。 */
  private readonly effectsInDeclarationOrder: readonly ActiveEffect[];

  constructor(operations: readonly ActiveEffect[]) {
    super();
    this.effectsInDeclarationOrder = operations;
  }

  apply(
    context: ReferenceContext,
    session: WorldSession,
    sameSlotSpawnSite: SameSlotSpawnSite | undefined,
  ): void {
    for (const operation of this.effectsInDeclarationOrder)
      operation.apply(context, session, sameSlotSpawnSite);
  }

  read(reader: EffectReader): void {
    for (const operation of this.effectsInDeclarationOrder) operation.read(reader);
  }

  /** 1つでも成立しない子があれば、合成も成立しない（並べた命令はすべて起こる約束のため）。 */
  override blocksOperation(context: ReferenceContext): boolean {
    return this.effectsInDeclarationOrder.some((operation) => operation.blocksOperation(context));
  }

  /** 子の合計。1つでも数えられない子（pick）があれば、合成も数えられない。 */
  override repeatLimitingVesselCount(): number | undefined {
    let total = 0;
    for (const operation of this.effectsInDeclarationOrder) {
      const vessels = operation.repeatLimitingVesselCount();
      if (vessels === undefined) return undefined;
      total += vessels;
    }
    return total;
  }

  /** 器を持つ子（ちょうど1つであることはロード時に確かめてある）に訊く。 */
  override acceptedCount(context: ReferenceContext, candidates: readonly WorldObject[]): number | undefined {
    for (const operation of this.effectsInDeclarationOrder) {
      const count = operation.acceptedCount(context, candidates);
      if (count !== undefined) return count;
    }
    return undefined;
  }
}

/**
 * set の1命令（対象プロパティへ絶対値を代入する、9.2節）。
 *
 * 値はリテラルか、**個体を1つ指す参照**（ObjectRef）。後者は解決した相手のインスタンスIDを書き込む
 * ——実行時に決まった個体を覚える手段はこれだけで、覚えた相手は `move` の行き先・`spawn` の配置先が
 * そのまま引く（筏が出航した海岸、docs/world/Voyage.md 3.5節）。
 */
export class SetEffect extends ActiveEffect {
  private readonly target: PropertyPath;
  private readonly value: number | ObjectRef;

  constructor(target: PropertyPath, value: number | ObjectRef) {
    super();
    this.target = target;
    this.value = value;
  }

  /** 参照が今どこも指していなければ何も書かない（解決できない適用は無視、9.1節）。 */
  apply(context: ReferenceContext): void {
    const value = typeof this.value === 'number' ? this.value : this.value.resolve(context)?.instanceId;
    if (value === undefined) return;
    this.target.propertyValue(context)?.setNumber(value);
  }

  read(reader: EffectReader): void {
    reader.set(this.target.root, this.target.propertyGlobalId, this.reading);
  }

  /** 代入する値の宣言（SetValueReading参照）。個体を指す参照は、指し方そのものを渡す。 */
  private get reading(): SetValueReading {
    return typeof this.value === 'number' ? this.value : this.value.reading;
  }
}

/** add の1命令（対象プロパティへ加減算する）。 */
export class AddEffect extends ActiveEffect {
  private readonly target: PropertyPath;
  private readonly amount: number;

  constructor(target: PropertyPath, amount: number) {
    super();
    this.target = target;
    this.amount = amount;
  }

  apply(context: ReferenceContext): void {
    this.applyScaled(context, 1, 1);
  }

  /**
   * transfer（9.5節）のlinked_add用: amount*numerator/denominatorにスケールした量を加減算する。
   * スケール後が0なら何もしない。
   */
  applyScaled(context: ReferenceContext, numerator: number, denominator: number): void {
    const scaled = (this.amount * numerator) / denominator;
    if (scaled === 0) return;
    this.target.propertyValue(context)?.add(scaled);
  }

  read(reader: EffectReader): void {
    reader.add(this.reading);
  }

  /** 自分を1件として名乗る読み上げ（AddReading参照）。transferのlinked_addもこの形で並ぶ。 */
  get reading(): AddReading {
    return {
      target: this.target.root,
      propertyGlobalId: this.target.propertyGlobalId,
      amount: this.amount,
    };
  }
}

/**
 * destroy の1命令（対象オブジェクトそのものを削除する、9.3節）。`destroy: [self, dragged]`は
 * 要素2つのDestroyEffectとして表す。same_slot spawnとの連携はsameSlotSpawnSite（ActiveEffect参照）が担う。
 *
 * **消し方に名前を添えられる**（reason、9.3節）。名前は消された側に残り、後から「どう消されたか」を
 * 答える（WorldObject.destroyedReason）——渇きで死んだのか、ただ立ち去ったのかは、消した宣言だけが知っている。
 */
export class DestroyEffect extends ActiveEffect {
  private readonly target: ObjectRef;

  /** この消滅の名前（9.3節）。書かれていなければundefinedで、消された側は何も名乗らない。 */
  private readonly reason: string | undefined;

  constructor(target: ObjectRef, reason?: string) {
    super();
    this.target = target;
    this.reason = reason;
  }

  apply(context: ReferenceContext): void {
    this.target.resolve(context)?.destroy(this.reason);
  }

  read(reader: EffectReader): void {
    reader.destroy(this.target.reading, this.reason);
  }
}

/**
 * spawn の配置先（9.4節）。スロットは指定せず、起点が持つスロットを宣言順に走査して最初に配置できた
 * 所へ入れる（著者がスロット名を知らなくてよい）。fallbackはYAML上に存在せず、配置失敗時は必ず起点
 * 自身の親へ伝播する（WorldObject.place参照）。
 *
 * **個体を指す形は `move` の移動先と同じ ObjectRef**（対象キー・プロパティが持つインスタンスID・型）。
 * 残る2つは個体ではないものを名乗るためにある——`same_slot` は位置、`child` は走査。
 */
export type SpawnTarget =
  /**
   * into 省略時の既定値。selfが今占めている場所（親と同じスロット）へ配置する。クラフト・腐敗など
   * 「同じ場所で別の物に置き換わる」場合に使う。一意の1スロットのため走査は行わない。
   */
  | 'same_slot'
  /**
   * selfの子を順に走査し、最初に受け取れた子のスロットへ入れる。intoが既に持つ「宣言順に走査して最初に
   * 配置できた所へ」という決め方が、1階層下へ伸びるだけ（docs/engine/TrapSystem.md 5.3節）。
   * 罠が自分の生んだ獲物へ怪我を渡す経路で、actorを持たないon_minからは他に手段が無い。
   */
  | 'child'
  /** 指した1つの個体が持つスロットを宣言順に走査する。 */
  | ObjectRef;

/**
 * spawn（9.4節）の1命令。intoへの配置に失敗した場合は起点の親へこぼれ、そこも受け取らなければさらに
 * 上へ遡る（WorldObject.spillTo）。**どの段でも枠の宣言はそのまま効く**ので、どこにも入らなければ
 * spawnしたオブジェクトは配置されないまま消える。
 */
export class SpawnEffect extends ActiveEffect {
  private readonly objectGlobalId: number;
  private readonly into: SpawnTarget;

  /**
   * 生む個数（9.4節、既定1）。同じ宣言を並べるのと同じ意味で、1個ずつ順に生んで配置する
   * ——「2個見つかる」を書くのに同じ行を2度書かせない。
   */
  private readonly count: number;

  constructor(objectGlobalId: number, into: SpawnTarget, count = 1) {
    super();
    if (!Number.isInteger(count) || count < 1)
      throw new Error(`countは1以上の整数である必要があります（値: ${count}）。`);

    this.objectGlobalId = objectGlobalId;
    this.into = into;
    this.count = count;
  }

  apply(
    context: ReferenceContext,
    session: WorldSession,
    sameSlotSpawnSite: SameSlotSpawnSite | undefined,
  ): void {
    for (let i = 0; i < this.count; i++)
      context.self?.executeSpawn(this.objectGlobalId, this.into, context, sameSlotSpawnSite);
  }

  read(reader: EffectReader): void {
    reader.spawn(this.objectGlobalId, this.count);
  }
}

/**
 * transfer（9.5節）の1命令。fromプロパティの実体値から、実際に出せる量とAmountの小さい方だけを
 * toプロパティへ移す（「在庫に応じて実際に動く量が変わる」移送）。YAMLはフラットな
 * `from`/`from_prop`/`to`/`to_prop`の4フィールドで表す。
 *
 * fromとtoの単位が違う場合（mLの水 → tick数のhydration）は、`amount`に対して受け取る側が
 * どれだけ増えるかを`to_amount`が持つ。**換算率をエンジンは知らず、移送する側が宣言する**。
 */
export class TransferEffect extends ActiveEffect {
  private readonly from: PropertyPath;
  private readonly to: PropertyPath;
  private readonly amount: number;
  private readonly toAmount: number;
  private readonly allowOverflow: boolean;
  private readonly linkedAdd: readonly AddEffect[];

  constructor(
    from: PropertyPath,
    to: PropertyPath,
    amount: number,
    allowOverflow: boolean,
    linkedAdd: readonly AddEffect[] = [],
    toAmount: number = amount,
  ) {
    super();
    // 受け取る量が0以下だと、出した分がどこにも入らない（9.5節）。
    if (toAmount <= 0) throw new Error(`'to_amount' は正の数である必要があります（値: ${toAmount}）。`);

    this.from = from;
    this.to = to;
    this.amount = amount;
    this.toAmount = toAmount;
    this.allowOverflow = allowOverflow;
    this.linkedAdd = linkedAdd;
  }

  /**
   * 出す量は「出せる量」（PropertyValue.availableToTransferOut）とAmountの小さい方。allow_overflowが
   * falseならさらに「受け取れる量」（remainingTransferCapacity）でも制限するが、単位が違えば
   * 受け取れる量は移送先の単位なので、**出す側の単位へ割り戻してから**比べる。
   *
   * linked_add（9.5節）は実際に出した量に比例（amount * actual_moved / Amount）してスケール適用する。
   * from/toが解決できない・対象がそのプロパティを持たない場合は何もしない。
   */
  apply(context: ReferenceContext): void {
    const fromValue: PropertyValue | undefined = this.from.propertyValue(context);
    const toValue: PropertyValue | undefined = this.to.propertyValue(context);
    if (fromValue === undefined || toValue === undefined) return;

    let taken = Math.min(this.amount, fromValue.availableToTransferOut());
    if (!this.allowOverflow) {
      const room = (toValue.remainingTransferCapacity() * this.amount) / this.toAmount;
      taken = Math.min(taken, room);
    }
    if (taken <= 0) return;

    fromValue.add(-taken);
    toValue.add((taken * this.toAmount) / this.amount);

    for (const linked of this.linkedAdd) linked.applyScaled(context, taken, this.amount);
  }

  /**
   * この輸送の宣言（TransferReading参照）。出す側は`amount`だけ減り、受け取る側は`to_amount`だけ
   * 増え、linked_addは全量移った場合の値で並ぶ。
   *
   * **在庫が満ちている前提の上限。** 実際に動く量は出せる量と空きで目減りする（applyがそれを見る）。
   */
  get reading(): TransferReading {
    return {
      from: this.from.root,
      fromPropertyGlobalId: this.from.propertyGlobalId,
      to: this.to.root,
      toPropertyGlobalId: this.to.propertyGlobalId,
      amount: this.amount,
      toAmount: this.toAmount,
      linked: this.linkedAdd.map((linked) => linked.reading),
    };
  }

  /**
   * 移送先が全candidatesで共通なら、その値が器になる（1つ）。移送先がdraggedなら、器はcandidateごとに
   * 別なので回数の上限を決めない（0）。
   */
  override repeatLimitingVesselCount(): number | undefined {
    return this.to.root === 'dragged' ? 0 : 1;
  }

  /**
   * 移送先の残り（PropertyValue.remainingTransferCapacity）が尽きるまでに、何個ぶん移せるか。
   * 移送先・移送元が解決できなければ、そこで打ち切る。
   */
  override acceptedCount(context: ReferenceContext, candidates: readonly WorldObject[]): number | undefined {
    if (this.to.root === 'dragged' || candidates.length === 0) return undefined;

    const toValue = this.to.propertyValue(context.withDragged(candidates[0]));
    if (toValue === undefined) return undefined;

    let room = toValue.remainingTransferCapacity();
    let count = 0;
    for (const candidate of candidates) {
      if (room <= 0) break;
      const fromValue = this.from.propertyValue(context.withDragged(candidate));
      if (fromValue === undefined) break;

      const taken = Math.min(this.amount, fromValue.availableToTransferOut());
      if (taken <= 0) break;
      room -= (taken * this.toAmount) / this.amount;
      count += 1;
    }
    return count;
  }

  read(reader: EffectReader): void {
    reader.transfer(this.reading);
  }

  /**
   * 輸送の両端を影響の辺として書き出す（PropertyInfluences参照）。
   *
   * **両端は互いの原因になる。** 受け取る側が増えるのは出す側の在庫があるからで、出す側が減るのは
   * 受け取る側へ持っていかれるから——1本の輸送が、どちらの端から見ても相手のせいで動いて見える。
   */
  collectTransferInfluences(declarer: WorldObject, active: boolean, out: InfluenceWriter): void {
    const from = declarer.resolveInfluenceTargets(this.from).at(0);
    const to = declarer.resolveInfluenceTargets(this.to).at(0);
    if (from === undefined || to === undefined) return;

    out.write({
      causeObject: from,
      causePropertyGlobalId: this.from.propertyGlobalId,
      target: to,
      targetPropertyGlobalId: this.to.propertyGlobalId,
      reversible: false,
      increases: true,
      active,
    });
    out.write({
      causeObject: to,
      causePropertyGlobalId: this.to.propertyGlobalId,
      target: from,
      targetPropertyGlobalId: this.from.propertyGlobalId,
      reversible: false,
      increases: false,
      active,
    });
  }
}
